client/src/main/java/io/pravega/client/stream/impl/Controller.java

```diff
@@ -86,17 +87,45 @@
     CompletableFuture<Boolean> deleteStream(final String scope, final String streamName);
 
     /**
-     * API to merge or split stream segments.
+     * API to request start of scale operation on controller. This method returns a future that will complete when
+     * controller service accepts the scale request.
+     *
+     * @param stream Stream object.
+     * @param sealedSegments List of segments to be sealed.
+     * @param newKeyRanges Key ranges after scaling the stream.
+     * @return A future which will throw if the operation fails, otherwise returning a boolean to
+     *         indicate that the scaling was started or not.
+     */
+    CompletableFuture<Boolean> startScale(final Stream stream, final List<Integer> sealedSegments,
+                                           final Map<Double, Double> newKeyRanges);
+
+    /**
+     * API to merge or split stream segments. This call returns a future that completes when either the scale
+     * operation is completed on controller service (succeeded or failed) or the specified timeout elapses.
      * 
      * @param stream Stream object.
      * @param sealedSegments List of segments to be sealed.
      * @param newKeyRanges Key ranges after scaling the stream.
+     * @param timeoutMillis timeout and throw exception if scale does not complete withing specified timeout.
+     * @param executorService executor to be used for busy waiting.
      * @return A future which will throw if the operation fails, otherwise returning a boolean to
-     *         indicate that the stream was scaled or false if the segments to seal have already
-     *         been sealed.
+     *         indicate that the scaling was completed successfully or not. It can also throw exception if the specified
+     *         timeout elapses.
      */
     CompletableFuture<Boolean> scaleStream(final Stream stream, final List<Integer> sealedSegments,
-                                           final Map<Double, Double> newKeyRanges);
+                                           final Map<Double, Double> newKeyRanges,
+                                           final long timeoutMillis,
+                                           final ScheduledExecutorService executorService);
```

Comment:
I don't like very much that we are adding a timeout and an executor here, it makes the API a lot uglier. I think we can get rid of the timeout, but I'm not so sure about the executor.